### PR TITLE
Draft: Rydberg simulator

### PR DESCRIPTION
**Pull request by fretchen**
### Summary
 This is an early draft for a Rydberg simulator as discussed in #64 .

### Details and comments
Concerning point 1. I am not sure how to implement `RX` and `RZ` from `qiskit-terra`. They should be extra gates similiar to `RLZ` and `RLX` ?
Concerning point 2. Not yet done, but fairly easy once point 3 to 5 work out.
Concerning point 3 and 4. They are working rather well in the present notebook.
Concerning point 5. This does not work yet. However, I do not understand the raised error at the end of the notebook. Any ideas there @eggerdj  or @LaurinFischer ?

Once, the problem with 5 is fixed and there is a clear idea on how to get 1, I think that it is fairly straight-forward...

### CODE REVIEW

**Comment by fretchen** [comment]
Hi @LaurinFischer. 

I was able to merge the fix into this PR draft to test a bit further. Indeed, the programs runs now. However, two problems remain here.

## Concerning point 1

I am not sure how to implement `RX` and `RZ` from `qiskit-terra`. The circuits are printed. But I cannot execute them locally or remotely. I am not sure if I should just overwrite them and wondered about possible feedback before I play around too much.

## Concerning point 5

It seems to work for two wires. However, there is a clear asymmetry in the locally simulated results for five wires. 
- This is not compatible with the experimental observations.
- This is also not compatible with the remote simulations.
- I had a look into the two four loops as it would have been my first guess that the indices are counted poorly. However, they look right.

Ideas for both ?

**Comment by LaurinFischer** [comment]
Hi @fretchen, 

Le'ts start the discussion on the qiskit `rx` and `rz` gates: You're right that currently, a `SpinBackend` from `qiskit-cold-atom` will not be able to parse them, and likewise, a qiskit-aer simulator will complain if you pass a circuit with an `rlz` and `rlx` gate, even though in the special case of spin=1/2 , they are essentially the same thing. 
However, I don't think that's necessarily a bad thing. The whole point of the circuits in `qiskit-cold-atom` is to have gates that are natively defined by their generating hamiltonians. This has no analogue in the standard qiskit package, so defining something like the `rydberg_full` gate in standard qiskit language is not really practical (maybe through the full parametrized unitary). 

In `qiskit-cold-atom`, setting the spin to s=1/2 (or any other value) doe not happen at the circuit level. The circuits themselves and the gates therein simply don't define the spin length, rather, it's the backend that sets the length of the spins. The `SpinSimulator` can even simulate the same circuit with different spin lenghts, that are passed as an option to the `backend.run()` method. 
Supporting standard qiskit gates would mean that we loose this flexibility on the circuit level. 

Therefor, I believe that keeping these two domains separated makes sense. A `qiskit-cold-atom` simulation should then only use `SpinGate` gates. What do you think @fretchen , @eggerdj ?

**Comment by fretchen** [comment]
I agree with the points you made. So I adapted the PR accordingly, added the rydberg blockade and now think that it is ready for review.

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 1]
```diff
@@ -0,0 +1,1088 @@
...
+    "    qc_rabi.assign_parameters(\n",
+    "        {omega_t: phase},\n",
+    "        inplace=False,\n",
+    "    )\n",
+    "    for phase in phases\n",
+    "]\n",
+    "\n",
+    "job_rabi = backend.run(rabi_list, shots=500)\n",
+    "result_rabi = job_rabi.result()\n",
+    "counts_rabi = job_rabi.result().get_counts()"
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:106]:
```suggestion
    "counts_rabi = result_rabi.get_counts()"
```

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 2]
```diff
@@ -0,0 +1,1088 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "034b74d1-81d8-4363-a8db-89af6424f442",
+   "metadata": {},
+   "source": [
+    "next we have to count the same of the two spins that are `1` at the same time"
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:114]:
```suggestion
    "next we have to count the sum of the probabilities of both spins being in state 1, respectively."
```

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 3]
```diff
@@ -0,0 +1,1088 @@
...
+    "\n",
+    "rabi_list = [\n",
+    "    qc_rabi.assign_parameters(\n",
+    "        {omega_t: phase},\n",
+    "        inplace=False,\n",
+    "    )\n",
+    "    for phase in phases\n",
+    "]\n",
+    "\n",
+    "job_rabi = backend.run(rabi_list, shots=500)\n",
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:104]:
```suggestion
    "n_shots = 500\n",
    "job_rabi = backend.run(rabi_list, shots=n_shots)\n",
```

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 4]
```diff
@@ -0,0 +1,1088 @@
...
+   "id": "7d05af5b-80cc-4490-aa9b-5e50a673dced",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "outcomes_rabi = [result_rabi.get_memory(i) for i in range(len(rabi_list))]\n",
+    "for i, outcome in enumerate(outcomes_rabi):\n",
+    "    for j, run in enumerate(outcome):\n",
+    "        outcomes_rabi[i][j] = np.array(run.split(' '), dtype = int)\n",
+    "outcomes_rabi = np.array(outcomes_rabi)"
+   ]
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:129]:
```suggestion
   "source": [
    "sum_rabi = []\n",
    "for i, _ in enumerate(phases):\n",
    "    counts_up = 0\n",
    "    for outcome, count in counts_rabi[i].items():\n",
    "        if outcome[0] == \"1\":\n",
    "            counts_up += count\n",
    "        if outcome[-1] == \"1\":\n",
    "            counts_up += count        \n",
    "    sum_rabi.append(counts_up/n_shots)"
   ]
```

I think this way to compute `sum_rabi` is a bit more direct and easier to parse than going through the memory like you're doing. Your method is also fine with me though.

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 5]
```diff
@@ -0,0 +1,1088 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "id": "59865b0c-de3b-499d-bf97-c189494d9847",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "means_rabi = outcomes_rabi.mean(axis=1)\n",
+    "sum_rabi = means_rabi.sum(axis=1)"
+   ]
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:140]: This cell is not needed if we use the above way to compute the `sum_rabi` directly.

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 6]
```diff
@@ -0,0 +1,1088 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "985b7f66-8e31-4440-b562-665198891ac5",
+   "metadata": {},
+   "source": [
+    "### Including the Blockade\n",
+    "\n",
+    "Next, we implement the protocol, which shows Rydberg blockade. This is done with the `rydberg_full` gate. It implements the already mentionned Hamiltonian time dynamics."
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:184]:
```suggestion
    "Next, we implement the protocol which shows Rydberg blockade. This is done with the `rydberg_full` gate. It implements the already mentioned Hamiltonian time dynamics."
```

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 7]
```diff
@@ -0,0 +1,1088 @@
...
+   "source": [
+    "Nwires = 2\n",
+    "backend = SpinSimulator()\n",
+    "qc_block = QuantumCircuit(QuantumRegister(Nwires, \"spin\"))\n",
+    "\n",
+    "all_modes=range(Nwires)\n",
+    "omega_t = Parameter(\"ω\")\n",
+    "phi_t = Parameter(\"φ\")\n",
+    "\n",
+    "qc_block.rydberg_full(omega = omega_t, delta =0, phi = phi_t, modes=all_modes)\n",
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:214]:
```suggestion
    "qc_block.rydberg_full(omega=omega_t, delta=0, phi=phi_t, modes=all_modes)\n",
```

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 8]
```diff
@@ -0,0 +1,1088 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "id": "393e9539-e248-487b-b1b1-e16b2b0feb84",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "phases = np.linspace(0, 2*np.pi,15)\n",
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:227]:
```suggestion
    "phases = np.linspace(0, 2*np.pi, 15)\n",
```

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 9]
```diff
@@ -0,0 +1,1088 @@
...
+    "\n",
+    "block_list = [\n",
+    "    qc_block.assign_parameters(\n",
+    "        {omega_t: phase, phi_t: phase*10},\n",
+    "        inplace=False,\n",
+    "    )\n",
+    "    for phase in phases\n",
+    "]\n",
+    "\n",
+    "job_block = backend.run(block_list, shots=500)\n",
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:237]:
```suggestion
    "job_block = backend.run(block_list, shots=n_shots)\n",
```

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 10]
```diff
@@ -0,0 +1,1088 @@
...
+    "result_block = job_block.result()"
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:238]:
```suggestion
    "result_block = job_block.result()\n",
    "counts_block = job_block.result().get_counts()"
```

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 11]
```diff
@@ -0,0 +1,1088 @@
...
+   "source": [
+    "outcomes = [result_block.get_memory(i) for i in range(len(block_list))]\n",
+    "for i, outcome in enumerate(outcomes):\n",
+    "    for j, run in enumerate(outcome):\n",
+    "        outcomes[i][j] = np.array(run.split(' '), dtype = int)\n",
+    "outcomes = np.array(outcomes)\n",
+    "\n",
+    "means_block = outcomes.mean(axis=1)\n",
+    "sum_block = means_block.sum(axis=1)"
+   ]
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:256]:
```suggestion
   "source": [
    "sum_block = []\n",
    "for i, _ in enumerate(phases):\n",
    "    counts_up = 0\n",
    "    for outcome, count in counts_block[i].items():\n",
    "        if outcome[0] == \"1\":\n",
    "            counts_up += count\n",
    "        if outcome[-1] == \"1\":\n",
    "            counts_up += count        \n",
    "    sum_block.append(counts_up/n_shots)"
   ]
```

Like above, I think this looks a bit more direct and simpler.

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 12]
```diff
@@ -0,0 +1,1088 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e5760868-6683-4c0f-a4f0-617159c67242",
+   "metadata": {},
+   "source": [
+    "now we can compare the dynamics without Rydberg blockade and with Rydberg blockade"
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:263]:
```suggestion
    "Now we can compare the dynamics without Rydberg blockade and with Rydberg blockade."
```

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 13]
```diff
@@ -0,0 +1,1088 @@
...
+    "ax.set_xlabel('phase')\n",
+    "ax.set_ylabel('sum of spins up')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a8184384-1239-4670-aaab-adb48cde349b",
+   "metadata": {},
+   "source": [
+    "We can see observe two things:\n",
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:309]:
```suggestion
    "We can observe two things:\n",
```

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 14]
```diff
@@ -0,0 +1,1088 @@
...
+    "\n",
+    "- The blockade almost completely suppressed the existence of both spins being up at the same time. This is due to its energetic cost.\n",
+    "- The oscillation frequency is increased by a factor of $\\sqrt(2)$. This is an indication of the coherent, entangled dynamics, which are discussed in detail in the original paper."
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:312]:
```suggestion
    "- The oscillation frequency is increased by a factor of $\sqrt{2}$. This is an indication of the coherent, entangled dynamics, which are discussed in detail in the original paper."
```

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 15]
```diff
@@ -0,0 +1,1088 @@
...
+   ],
+   "source": [
+    "Nwires = 5\n",
+    "backend = SpinSimulator()\n",
+    "qc_rabi = QuantumCircuit(QuantumRegister(Nwires, \"spin\"))\n",
+    "\n",
+    "all_modes=range(Nwires)\n",
+    "omega_t = Parameter(\"ω\")\n",
+    "\n",
+    "qc_rabi.rlx( omega_t, all_modes)\n",
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:424]:
```suggestion
    "qc_rabi.rlx(omega_t, all_modes)\n",
```

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 16]
```diff
@@ -0,0 +1,1088 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "id": "39c035f3-0af9-4ed7-9ab9-47c836a96827",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "phases = np.linspace(0, 2*np.pi,15)\n",
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:437]:
```suggestion
    "phases = np.linspace(0, 2*np.pi, 15)\n",
```

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 17]
```diff
@@ -0,0 +1,1088 @@
...
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "means_rabi = outcomes_rabi.mean(axis=1)\n",
+    "\n",
+    "f, ax = plt.subplots()\n",
+    "ax.pcolormesh(phases,np.arange(Nwires), means_rabi.T)\n",
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:488]:
```suggestion
    "im = ax.pcolormesh(phases, np.arange(Nwires), means_rabi.T)\n",
    "f.colorbar(im, ax=ax)\n",
```

just to add a colorbar

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 18]
```diff
@@ -0,0 +1,1088 @@
...
+   "source": [
+    "Nwires = 7\n",
+    "backend = SpinSimulator()\n",
+    "qc_block = QuantumCircuit(QuantumRegister(Nwires, \"spin\"))\n",
+    "\n",
+    "all_modes=range(Nwires)\n",
+    "omega_t = Parameter(\"ω\")\n",
+    "phi_t = Parameter(\"φ\")\n",
+    "\n",
+    "qc_block.rydberg_full(omega = omega_t, delta =0, phi = phi_t, modes=all_modes)\n",
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:528]:
```suggestion
    "qc_block.rydberg_full(omega=omega_t, delta=0, phi=phi_t, modes=all_modes)\n",
```

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 19]
```diff
@@ -0,0 +1,1088 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "id": "4b9d766c-6965-450a-ac88-721927823432",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "phases = np.linspace(0, 2*np.pi,15)\n",
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:541]:
```suggestion
    "phases = np.linspace(0, 2*np.pi, 15)\n",
```

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 20]
```diff
@@ -0,0 +1,1088 @@
...
+       "<Figure size 640x480 with 1 Axes>"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "f, ax = plt.subplots()\n",
+    "ax.pcolormesh(phases,np.arange(Nwires), means_block.T)\n",
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:600]:
```suggestion
    "im = ax.pcolormesh(phases, np.arange(Nwires), means_block.T)\n",
    "f.colorbar(im, ax=ax)\n",
```

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 21]
```diff
@@ -0,0 +1,1088 @@
...
+    "ax.set_ylabel('site')\n",
+    "ax.set_xlabel('time')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7a1a1b5e-054a-4891-a19e-23d37debef7a",
+   "metadata": {},
+   "source": [
+    "As you can see the interactions now substantially altered the dynamics:\n",
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:610]:
```suggestion
    "The interactions now substantially altered the dynamics:\n",
```

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 22]
```diff
@@ -0,0 +1,1088 @@
...
+    "\n",
+    "- On the outer sites that first oscillation appears quicker, which is reminiscent of the blockade scenario with two qubits.\n",
+    "- On the site 1 and 5 the dynamics is strongly suppressed due to the strong interaction with all the neighbors.\n",
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:613]:
```suggestion
    "- On site 1 and 5 the dynamics are strongly suppressed due to the strong interaction with all the neighbors.\n",
```

File: docs/tutorials/09_ryberg_dynamics.ipynb [hunk 23]
```diff
@@ -0,0 +1,1088 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "993f6b14-972e-4239-8de6-8efee5f3a74c",
+   "metadata": {},
+   "source": [
+    "## Optional: Working on a cloud simulator\n",
+    "\n",
+    "We would also like to send such circuits to some real hardware and investigate these results. In this last section, we will present this possibility with the [alqor](https://qlued.alqor.io) provider. \n",
```
Comment (LaurinFischer) [inline docs/tutorials/09_ryberg_dynamics.ipynb:624]: Could you rephrase the term "real hardware". It sounds a bit misleading.

**Review by LaurinFischer — COMMENTED** [review]
Hi Fred, 

this looks already very good and will make for a nice demonstration to add to our tutorials, thanks a lot! My comments are mostly nitpicks on code style and language to make the notebook a bit cleaner. 

Also, we should add a releasenote that summarizes the the changes made by this PR.

**Comment by fretchen** [comment]
I think that I took into account all comments. 

The only one that involved coding was the part on the tests. There I added specific tests for `omega` and `delta` non-zero to make sure that an analytical expression is possible for each case.

**Comment by eggerdj** [comment]
Thanks @fretchen This looks really cool. So far I am reading through the tutorial notebook. It's very nice. One aspect that can IMO be made a bit better is the connection of the Rydberg blockade to the Hamiltonian term. E.g. we could change this sentence
```
Next, we implement the protocol which shows Rydberg blockade. 
This is done with the rydberg_full gate. 
It implements the already mentioned Hamiltonian time dynamics.
```
to
```
Next, we implement the protocol which shows Rydberg blockade. 
This is done with the rydberg_full gate. It implements the already mentioned Hamiltonian time dynamics. 
The last term including $\hat n_i\hat n_j$ is responsible for the Rydberg blockade which can, 
e.g., energetically penalizes situations where both spins are up. 
```
In cell 7: you have the lowercase greek letters for the Hamiltonian but the math shows the upper case letters. Can we make them match?

By the way, why do you chose `phi` and `10*phi` in cell 8? Is this to follow the experimental setup in the reference? Perhaps worth a comment.

In the controlled entanglement section:
```
Let us now also visualized in a single circuit how the rydberg Blockade might be used to generate an entangled state. 
For this we will employ the gate rydberg_block, which generates the blockade.
```
Here, `rydberg_block` applies the `n_i n_j` term correct? Can we perhaps explicitly write this? 

In cell 12: this is a Bell state correct? Perhaps write it explicitly?

Can we have a few more sentences in the introduction to `Spin structures from Rydberg interactions`? E.g. what is meant by `larger spin structures`? Also change `The interactions now substantially altered the dynamics:` to `The Rydberg blockade interactions now substantially altered the dynamics:`

The notebook is sort of missing a conclusion. Can you add one before the optional section? E.g. something like summarizing what we have learnt?

Anyhow the tutorial is very nice and I'm being a bit picky with my comments above.

File: qiskit_cold_atom/spins/spins_gate_library.py [hunk 24]
```diff
@@ -371,3 +371,167 @@ def __init__(self, num_atoms: int):
...
+        # add generators of detuning
+        if delta != 0.0:
+            for i in range(self.num_modes):
+                generators.append((f"Z_{i}", delta))
+
+        # add generators of interaction term
+        if phi != 0.0:
+            for i in range(self.num_modes):
+                for j in range(i + 1, self.num_modes):
+                    coeff = phi / np.abs(i - j) ** 6
```
Comment (eggerdj) [inline qiskit_cold_atom/spins/spins_gate_library.py:448]: Here, it looks like you have made an assumption on the topology of the device: linear connectivity. I.e. the spin ordering and `1/abs(i - j) ** 6` is only valid for a linear arrangement of the spins. If I now have a 2D array of atoms (or any other connectivity) this no longer holds. Correct? At the very least we need to document this limitation. Ideally we should come up with a design that is more flexible. I see a few potential ways out:
* Let the user provide the coeff(s): requires more work from the user. Perhaps not the best option.
* Somehow bring in the coupling map. Perhaps we can leverage `qiskit_nature.second_q.hamiltonians.lattices`?
In any case I think it would proceed as follows: (i) document the limitation in this PR (ii) open an issue to discuss potential solutions.

File: test/test_spin_gates.py [hunk 25]
```diff
@@ -179,6 +181,118 @@ def test_lz2_gate(self):
...
+    def test_rydberg_block_gate(self):
+        """check matrix form of the rydberg blockade gate on two qubits"""
+        chi = 2 * np.pi
+        circ = QuantumCircuit(2)
+        circ.append(RydbergBlockade(2, phi=chi), qargs=[0, 1])
+        # add gate to circuit via the @add_gate-decorated method
+        circ_decorated = QuantumCircuit(2)
+        circ_decorated.rydberg_block(chi, [0, 1])
+        for circuit in [circ, circ_decorated]:
+            unitary = self.backend.run(circuit).result().get_unitary()
```
Comment (eggerdj) [inline test/test_spin_gates.py:193]: I did not know we could do this. If we are just testing the `RydbergBlockade` instruction could we do:
```python
from qiskit.quantum_info import Operator

np.allclose(Operator(RydbergBlockade(2, phi=chi)).data, expected)
```
Does that work?

Comment (eggerdj) [reply test/test_spin_gates.py:193]: That being said, I also like how the decorated circuit is being tested too. Perhaps we can have an additional test using the Quantum Info. operator too? Maybe with different parameters?

**Review by eggerdj — CHANGES_REQUESTED** [review]
This looks good to me. The tutorial is nice. The one limitation I see is the assumption on the topology that needs to be discussed. Can you include a reno?

File: qiskit_cold_atom/spins/spins_gate_library.py (hunk 24 above)
Comment (fretchen) [reply qiskit_cold_atom/spins/spins_gate_library.py:448]: Agreed with the proposition. I will add some documentation of the limitations.

File: qiskit_cold_atom/spins/spin_circuit_solver.py [hunk 26]
```diff
@@ -92,21 +92,15 @@ def _embed_operator(self, operator: SpinOp, num_wires: int, qargs: List[int]) ->
         if not isinstance(operator, SpinOp):
             raise QiskitColdAtomError(f"Expected SpinOp; got {type(operator).__name__} instead")
 
-        embedded_op_list = []
```
[the remark this answers is not in the document]
Comment (fretchen) [reply qiskit_cold_atom/spins/spin_circuit_solver.py:95]: Exactly.

File: test/test_spin_gates.py (hunk 25 above)
Comment (fretchen) [reply test/test_spin_gates.py]: I think hat I was able to implement your suggestion.

**Comment by fretchen** [comment]
@eggerdj I think that it was fairly straight forward to implement all the changes that you suggested.

**Review by eggerdj — APPROVED** [review]
LGTM! Thanks a lot.
